<a href="https://colab.research.google.com/github/TechnoDexx/DataScience/blob/main/%D0%9F%D0%BE%D0%B3%D1%80%D1%83%D0%B6%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B2_DS%2C_%D0%B4%D0%B5%D0%BD%D1%8C_2_%D1%80%D0%B5%D1%88%D0%B5%D0%BD%D0%B8%D0%B5_%D0%94%D0%97_%D0%9B%D0%B5%D0%B1%D0%B5%D0%B4%D0%B5%D0%B2%D0%BE%D0%B9_%D0%94%D0%B0%D1%80%D1%8C%D0%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

# считываем данные по ссылке
data = pd.read_csv('https://raw.githubusercontent.com/MariaZharova/3-day-intensive/main/Mall_Customers.csv', sep = ',')

In [ ]:
# библиотеки для машинного обучения
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn import metrics

# сразу сделаем небольшое преобразование - переведём пол из строк в числа
data.Genre = data.Genre.apply(lambda x: 0 if x == 'Male' else 1)

# выделим для удобства таргет и признаки
X = data.drop('Score', axis = 1) # характеристики
y = data.Score # таргеты (правильные ответы)

# делим на данные для обучения и для теста
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### ⛳ Домашнее задание


✅ Задание 1:  Назовите число колонок в файле

In [ ]:
print('Число колонок в файле =', data.shape[1])

Число колонок в файле = 5


✅ Задание 2:  Назовите имя самой первой колонки

In [ ]:
print('Имя самой первой колонки:', data.columns[0])

Имя самой первой колонки: CustomerID


✅ Задание 3:  Найдите самую масимальную оценку от пользователя

In [ ]:
print('Самая масимальная оценка от пользователя =', data.Score.max())

Самая масимальная оценка от пользователя = 99


✅ Задание 4*: Проверьте, есть ли в данных дубликаты, в ответ напишите "да" или "нет"

*Подсказка: попробуйте использовать метод drop_duplicates

In [ ]:
data_new = data.copy()
data_new.drop_duplicates()
print('Нет') if data.shape[0] == data_new.shape[0] else print('Да')

Нет


✅ Задание 5: Попробуйте найти в документации ещё какие-нибудь модели из sklearn, и по аналогии с последней ячейкой обучить их на наших данных. Количество моделей неограничено:) В качестве итога можете посчитать метрику score и выбрать наилучшую модель.

Подсказка: можно поменять только название модели в первой строке, весь остальной код будет аналогичным.

- Цель: написать алгоритм (модель машинного обучения), который бы как можно точнее предсказывал Target, видя только столбцы-характеристики. Например, очень популярный подход - подобрать такие коэффициенты $a_1, a_2, ..., a_N$, чтобы сумма $x_1\cdot a_1 + ... + x_N\cdot a_N$ максимально точно приближала истинный ответ Target (пытаемся вывести закономерности, понять, какие столбцы влияют на итоговый ответ больше всего).

In [ ]:
# 1 вариант кода с другой моделью: применяем "Случайный лес" - RandomForestRegressor
# RandomForestRegressor - тип метода ансамблевого обучения для регрессии
# RandomForestRegressor работает путем построения нескольких деревьев решений во время обучения и вывода среднего прогноза для отдельных деревьев

from sklearn.ensemble import RandomForestRegressor

reg = RandomForestRegressor() # 1
reg.fit(X_train, y_train) # 2
y_pred = reg.predict(X_test) # 3
rmse = (np.sqrt(metrics.mean_squared_error(y_test, y_pred))) # 4

print(f"The model performance for testing set") # красивая печать на экран :)
print(f"--------------------------------------")
print(f"Root Mean Squared Error: {rmse}")

The model performance for testing set
--------------------------------------
Root Mean Squared Error: 20.043671195666725


In [ ]:
# 2 вариант кода с другой моделью: применяем Метод опорных векторов - SVR
# SVR - тип алгоритма регрессии, который использует машины опорных векторов для выполнения задач регрессии
# SVR работает путем поиска гиперплоскости, которая максимизирует разницу между прогнозируемыми значениями и фактическими значениями

from sklearn.svm import SVR

reg = SVR() # 1
reg.fit(X_train, y_train) # 2
y_pred = reg.predict(X_test) # 3
rmse = (np.sqrt(metrics.mean_squared_error(y_test, y_pred))) # 4

print(f"The model performance for testing set") # красивая печать на экран :)
print(f"--------------------------------------")
print(f"Root Mean Squared Error: {rmse}")

The model performance for testing set
--------------------------------------
Root Mean Squared Error: 24.51660326402299


In [ ]:
# 3 вариант кода с другой моделью: применяем Метод K-ближайших соседей - KNeighborsRegressor
# KNeighborsRegressor - тип алгоритма обучения на основе экземпляров
# KNeighborsRegressor работает путем поиска k-ближайших соседей к заданной точке данных
# и использования их среднего значения в качестве прогнозируемого значения для точки данных

from sklearn.neighbors import KNeighborsRegressor

reg = KNeighborsRegressor() # 1
reg.fit(X_train, y_train) # 2
y_pred = reg.predict(X_test) # 3
rmse = (np.sqrt(metrics.mean_squared_error(y_test, y_pred))) # 4

print(f"The model performance for testing set") # красивая печать на экран :)
print(f"--------------------------------------")
print(f"Root Mean Squared Error: {rmse}")

The model performance for testing set
--------------------------------------
Root Mean Squared Error: 19.543029447861965


# 🔎 Выводы:
Если показатель Root Mean Squared Error (RMSE) меньше, это означает, что модель работает лучше и точнее. RMSE является метрикой, которая показывает разность между предсказанными и наблюдаемыми значениями в модели.

Модель линейной регрессии - LinearRegression даёт Среднеквадратичную ошибку, т.е. Root Mean Squared Error = 22.46448986768789

Модель "Случайный лес" - RandomForestRegressor даёт Среднеквадратичную ошибку, т.е. Root Mean Squared Error = 20.464614154681733

Метод опорных векторов - SVR даёт Среднеквадратичную ошибку, т.е. Root Mean Squared Error = 24.51660326402299

Метод K-ближайших соседей - KNeighborsRegressor даёт Среднеквадратичную ошибку, т.е. Root Mean Squared Error = 19.543029447861965

Таким образом, Метод K-ближайших соседей - KNeighborsRegressor даёт наименьшее значение из всех опробованных методов, т.е. он показал лучший результат.
Более низкое значение RMSE указывает на то, что модель имеет меньшую среднеквадратичную ошибку и лучше соответствует данным.

In [ ]:
# Исходный вариант кода от преподавателя: применяется модель линейной регрессии - LinearRegression
from sklearn.linear_model import LinearRegression

reg = LinearRegression() # 1
reg.fit(X_train, y_train) # 2
y_pred = reg.predict(X_test) # 3
rmse = (np.sqrt(metrics.mean_squared_error(y_test, y_pred))) # 4

print(f"The model performance for testing set") # красивая печать на экран :)
print(f"--------------------------------------")
print(f"Root Mean Squared Error: {rmse}")

The model performance for testing set
--------------------------------------
Root Mean Squared Error: 22.46448986768789


✅ Задание 6*: Попробуйте сделать предсказание на своих собственных данных. Можете воспользоваться приведённым шаблоном

In [ ]:
import numpy as np

# генерируем собственные случайные данные для 10 покупателей: ID (в диапазоне от 1 до 1 млн.)
# пола (0 - жен, 1 - муж), возраста (от 18 до 100 лет включительно) и суммы покупки (от 1 до 100 тыс. рублей)

id_list = np.random.randint(1, 1_000_001, size=10)
gender_list = np.random.randint(0, 2, size=10)
age_list = np.random.randint(18, 101, size=10)
purchase_list = np.random.randint(1, 100_001, size=10)

# объединяем сгенерированные списки в массив numpy
data = np.column_stack((id_list, gender_list, age_list, purchase_list))

print(data)
print()

# создаем модель и загружаем в нее веса
reg = LinearRegression()
reg.fit(X_train, y_train)

# делаем предсказания на своих данных
data_to_predict = np.array([1, 0, 20, 50_000]).reshape(1, -1)
print("Предсказание на собственных данных:", reg.predict(data_to_predict))

[[587134      1     44  29568]
 [316213      1     94  74468]
 [ 36086      1     45  29071]
 [ 66474      1     39   1997]
 [389535      1     76  65931]
 [849136      1     93  66758]
 [503628      1     33  78260]
 [176858      1     82  39292]
 [117682      1     63  71698]
 [442006      0     33  42902]]

Предсказание на собственных данных: [-10051.30536044]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(



🌟 PS: Можете подумать, какая колонка является неинформативной с точки зрения обучения (соответственно, её можно удалить:))



# 🌟 Мой ответ: **неинформативной с точки зрения обучения является колонка CustomerID или ID**